src/ansi-c/goto-conversion/goto_clean_expr.cpp

```diff
@@ -440,9 +476,155 @@ goto_convertt::clean_expr_resultt goto_convertt::clean_expr(
   }
   else if(expr.id() == ID_forall || expr.id() == ID_exists)
   {
+    quantifier_exprt &qex = to_quantifier_expr(expr);
+    code_expressiont code{qex.where()};
     DATA_INVARIANT(
-      !has_subexpr(expr, ID_side_effect),
-      "the front-end should check quantified expressions for side-effects");
+      !has_subexpr(expr, ID_side_effect) ||
+        (code.operands()[0].id() == ID_side_effect &&
+         code.operands()[0].get_named_sub()[ID_statement].id() ==
+           ID_statement_expression),
+      "quantifier must not contain side effects");
+
+    // Handle the case that quantifier body is a statement expression.
+    if(
+      code.operands()[0].id() == ID_side_effect &&
+      code.operands()[0].get_named_sub()[ID_statement].id() ==
+        ID_statement_expression)
+    {
+      goto_programt where;
+      goto_convert(code, where, mode);
+      where.compute_location_numbers();
+
+      natural_loops_mutablet natural_loops(where);
+      INVARIANT(
+        natural_loops.loop_map.size() == 0,
+        "quantifier must not contain side effects");
+
+      // `last` is the instruction corresponding to the last expression in the
+      // statement expression.
+      goto_programt::const_targett last;
+      for(goto_programt::const_targett it = where.instructions.begin();
+          it != where.instructions.end();
+          ++it)
+      {
+        // `last` is an other-instruction.
+        if(it->is_other())
+        {
+          last = it;
+        }
+      }
+      auto last_expr =
+        expr_try_dynamic_cast<code_expressiont>(last->get_other())
+          ->expression();
+
+      // `pathes` contains all the `targett` we are iterating over.
+      std::vector<goto_programt::const_targett> pathes;
+      pathes.push_back(where.instructions.begin());
+      std::vector<std::pair<exprt, replace_symbolt>>
```

Comment:
Is it the case that this has to have exactly the same number of entries as `pathes`? Perhaps you want to introduce a local `struct` that has three members and then just have a single `std::vector` that holds elements of that struct type?